docs/proposals/20200304-autoscaler-scale-ux.md

```diff
@@ -0,0 +1,117 @@
+the ability to scale from zero for a given machineset or machinedeployment.
+
+## Proposal Ideas
+
+There are several proposals to consider.  Once we decide on one, we'll eliminate
+the others.
+
+### Annotations
+Modify the autoscaler to look at particular annotations on a given object to determine the appropriate amounts of each resource.
+
+#### Benefits
+Easy, simple.  Users can modify if needed (we can instruct controllers to only
+add annotations if they're missing, etc), users can bring their own clouds
+and annotate their objects to support scale from zero and we don't need to worry
+about supporting every cloud day 1.
+
+#### Cons
+We're essentially creating an API around an annotation, and the api folks don't
+like that type of thing.
+
+### Spec Field
+We can add fields to the spec of a particular object to set these items.
+
+#### Benefits
+Well defined API.
+
+Users can set these fields, or some controller can
+populate them.
+
+#### Cons
+Automatically populating Spec fields is not ideal.  For starters, a user might
+copy an existing machineset to create a new one with a different instance type,
+and might overlook these fields because they didn't set them.  This would
+definitely cause problems.
+
+What do we do if the user specifies some fields and not others.
+
+Automatically populating spec seems to violate 'spec == user intent' principle.
+
+### Status Field
+We can add fields to the status of an object.
+
+#### Benefits
+Makes the most sense.  If it's information we're determining programatically
+about an object that other things need to consume, it should probably go here.
+
+Well defined API.
+
+#### Cons
+Users can't easily update this field if they want to set values themselves for
+unsupported clouds or clouds where billing data is not common or available
+such as OpenStack.
+
+### Status Field + Override Mechanism
+Use the status field method as mentioned above, but add an override field to
+the Spec or support an annotation on the machineset/machinedeployment object
+to copy information into the status field.  The autoscaler always reads from
+the status field, we're just making it easy to get data there.
+
+#### Benefits
+All the benefits of the Status Field option above. Users can set their own
+values easily without having to create special status client to do it (eg,
+can use kubectl or specify the values at object creation time).
+
+Since we're calling these new override fields (either in the spec or annoation)
+it should be clear what the intent of the values are.  It solves the problem
+of user only specifies some fields and not others (eg, they only override CPU).
+
+Clean separation of discovered values and user intent.
+
+#### Cons
+We have to build something that copies these values into the staus.  This seems
+like it could be an easily shared piece of library code.
+
+Multiple controllers updating the same field on the status object, need to
```

Comment:
I'd like to avoid the potential for conflicting controllers attempting to manage the same field in a resource's Status field.

We also need to keep in mind that the controller for the resource "owns" the status of that resource. Ideally we'd be able to copy these values from a resource that is owned by the controller that is generating them.

Reply:
There's not really any place to copy the information from.  Unless we create some sort of new 'scale record' or some crazy thing like that, and then that's yet-another-crd.